File: cs/Evaluator/Evaluator.cs
global using System;
global using System.Collections.Generic;
using Model;
using System.Linq;

namespace Evaluator;

internal record struct Context(int[] Param)
{
    internal Context(int[] param, ExternalSource simpleExternal, ExternalSource uniqueExternal, ExternalSource orderedExternal)
        : this(param)
    {
        SimpleExternal = simpleExternal;
        UniqueExternal = uniqueExternal;
        OrderedExternal = orderedExternal;
    }

    internal readonly ExternalSource SimpleExternal;
    internal readonly ExternalSource UniqueExternal;
    internal readonly ExternalSource OrderedExternal;
}

/// <summary>Classe permettant de simuler l'execution du code</summary>
public static class Evaluator
{
    /// <summary>Execute l'application donnée en paramètre</summary>
    /// <param name="app">L'application a executer</param>
    public static int Run(Application app)
        => Run(app, new DictionarySource(new()), new DictionarySource(new()), new ConsoleSource());

    /// <summary>Execute l'application donnée en paramètre</summary>
    /// <param name="app">L'application a executer</param>
    /// <param name="simpleExternal">La valeur des simple external</param>
    /// <param name="uniqueExternal">La valeur des unique external</param>
    /// <param name="orderedExternal">La valeur des ordered external</param>
    public static int Run(Application app, ExternalSource simpleExternal, ExternalSource uniqueExternal, ExternalSource orderedExternal)
        => Evaluate(app.Main, new Context(Array.Empty<int>(), simpleExternal, uniqueExternal, orderedExternal));

    private static int Evaluate(Fonction f, Context ctx)
    {
        Dictionary<Expression, int> valeurs = new();
        int current = 0;
        foreach (Expression item in f.Actions)
        {
            current = Evaluate(item, valeurs, ctx);
            valeurs[item] = current;
        }

        return current;
    }

    private static int Evaluate(Expression expr, Dictionary<Expression, int> valeurs, Context ctx)
    {
        return expr switch
        {
            AdditionExpression ae => valeurs[ae.First] + valeurs[ae.Second],
            ConditionalExpression ce => valeurs[ce.Cond] == 0
                ? Evaluate(ce.ZeroBranch, valeurs, ctx)
                : Evaluate(ce.NonZeroBranc, valeurs, ctx),
            ConstExpression ce => ce.Valeur,
            SimpleExternal se => ctx.SimpleExternal.GetValue(se.Valeur),
            UniqueExternal ue => ctx.UniqueExternal.GetValue(ue.Valeur),
            OrderedExternal oe => ctx.OrderedExternal.GetValue(oe.Valeur),
            FuncCall fc
            => Evaluate(fc.Func, ctx with { Param = fc.Param.Select(item => valeurs[item]).ToArray() }),
            ParamExpression pe => ctx.Param[pe.Valeur],
            _ => throw new Exception(),
        };
    }
}

File: hs/src/Evaluator/Evaluator.hs
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE FlexibleContexts #-}

module Evaluator.Evaluator
  ( eval,
    evalParam,
    run,
  )
where

import Control.Monad.Except (MonadError (throwError))
import Evaluator.EvaluationContext
import Evaluator.Helper (Error (..), readValue)
import Model.Model (Application (..), Expression (..), Fonction (..), InstructionList (..), SimpleExpression (..))
import Text.Read (readMaybe)

run :: ExceptT Error IO Int -> IO ()
run main =
  runExceptT main >>= internalRun

internalRun :: Either Error Int -> IO ()
internalRun (Left (Error e)) = putStrLn e
internalRun (Right program) = putStrLn ("Resultat : " ++ show program)

eval :: EvalContext m => Application -> m Int
eval app = evalParam app []

evalParam :: EvalContext m => Application -> [Int] -> m Int
evalParam (Application mainFunc otherFunc) =
  runFonction mainFunc (GlobalContext otherFunc 0)

-- Comment supprimer cette fonction
tmpInit :: (ParamContext c) => Int -> c -> RunContext c
tmpInit = initEvaluationContext

runFonction :: (EvalContext m, FonctionContext c) => Fonction -> c -> [Int] -> m Int
runFonction (Fonction (InstructionList first others) paramCount) context funcParam
  | Prelude.length funcParam /= paramCount = throwError $ Error "Invalid number of call arguments"
  | otherwise =
    let preFuncContext = CallContext funcParam context
     in do
          firstValue <- evalExpressionWithPreContext first preFuncContext
          let tmp = tmpInit firstValue preFuncContext
          evalExpressions others tmp

type EvalContext m = (MonadError Error m, MonadIO m)

evalExpressions :: (EvalContext m, EvaluationContext c) => [Expression] -> c -> m Int
evalExpressions (current : nexts) context =
  do
    val <- evalExpressionInContext context current
    evalExpressions nexts val
evalExpressions [] context = return $ head $ stack context

evalExpressionInContext :: (EvalContext m, EvaluationContext c) => c -> Expression -> m c
evalExpressionInContext context expr =
  do
    newVal <- evalExpression expr context
    return $ evolveContext context newVal

evalExpression :: (EvalContext m, EvaluationContext c) => Expression -> c -> m Int
evalExpression (AdditionExpression first second) context =
  do
    param1 <- readContext context first
    param2 <- readContext context second
    return $ param1 + param2
--
evalExpression (ConditionalExpression cond notNullExpression nullExpression) context =
  let expressionSelector x = if x == 0 then nullExpression else notNullExpression
   in readContext context cond >>= (\x -> evalExpression (expressionSelector x) context)
--
evalExpression (FuncCall funcIndex params) context =
  do
    func <- readFuncInContext context funcIndex
    param <- mapM (readContext context) params
    runFonction func context (toList param)
--
evalExpression (Simple expr) context =
  evalExpressionWithPreContext expr context

evalExpressionWithPreContext :: (EvalContext m, ParamContext c) => SimpleExpression -> c -> m Int
evalExpressionWithPreContext InvalidExpression _ = throwError (Error "Invalid Expression")
--
evalExpressionWithPreContext (ConstExpression val) _ = return val
--
evalExpressionWithPreContext (ParamExpression paramIndex) context =
  readParamInContext context paramIndex
--
evalExpressionWithPreContext (SimpleCall funcIndex) context =
  readFuncInContext context funcIndex >>= (\x -> runFonction x context [])
--
evalExpressionWithPreContext (ExternalExpression _ _) _ = liftIO readValue
